source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/direct/factory/factory_env.py

```diff
@@ -0,0 +1,890 @@
+
+            # Simulate and update tensors.
+            self.step_sim_no_action()
+            ik_time += self.physics_dt
+
+        return pos_error, axis_angle_error
+
+    def get_handheld_asset_relative_pose(self):
+        """Get default relative pose between help asset and fingertip."""
+        if self.cfg_task.name == "peg_insert":
+            held_asset_relative_pos = torch.zeros_like(self.held_base_pos_local)
+            held_asset_relative_pos[:, 2] = self.cfg_task.held_asset_cfg.height
+            held_asset_relative_pos[:, 2] -= self.cfg_task.robot_cfg.franka_fingerpad_length
+        elif self.cfg_task.name == "gear_mesh":
+            held_asset_relative_pos = torch.zeros_like(self.held_base_pos_local)
+            gear_base_offset = self._get_target_gear_base_offset()
+            held_asset_relative_pos[:, 0] += gear_base_offset[0]
+            held_asset_relative_pos[:, 2] += gear_base_offset[2]
+            held_asset_relative_pos[:, 2] += self.cfg_task.held_asset_cfg.height / 2.0 * 1.1
+        elif self.cfg_task.name == "nut_thread":
+            held_asset_relative_pos = self.held_base_pos_local
+        else:
+            raise NotImplementedError("Task not implemented")
+
+        held_asset_relative_quat = self.identity_quat
+        if self.cfg_task.name == "nut_thread":
+            # Rotate along z-axis of frame for default position.
+            initial_rot_deg = self.cfg_task.held_asset_rot_init
+            rot_yaw_euler = torch.tensor([0.0, 0.0, initial_rot_deg * np.pi / 180.0], device=self.device).repeat(
+                self.num_envs, 1
+            )
+            held_asset_relative_quat = torch_utils.quat_from_euler_xyz(
+                roll=rot_yaw_euler[:, 0], pitch=rot_yaw_euler[:, 1], yaw=rot_yaw_euler[:, 2]
+            )
+
+        return held_asset_relative_pos, held_asset_relative_quat
+
+    def _set_franka_to_default_pose(self, joints, env_ids):
+        """Return Franka to its default joint position."""
+        gripper_width = self.cfg_task.held_asset_cfg.diameter / 2 * 1.25
+        joint_pos = self._robot.data.default_joint_pos[env_ids]
+        joint_pos[:, 7:] = gripper_width  # MIMIC
+        joint_pos[:, :7] = torch.tensor(joints, device=self.device)[None, :]
+        joint_vel = torch.zeros_like(joint_pos)
+        joint_effort = torch.zeros_like(joint_pos)
+        self.ctrl_target_joint_pos[env_ids, :] = joint_pos
+        print(f"Resetting {len(env_ids)} envs...")
+        self._robot.set_joint_position_target(self.ctrl_target_joint_pos[env_ids], env_ids=env_ids)
+        self._robot.write_joint_state_to_sim(joint_pos, joint_vel, env_ids=env_ids)
+        self._robot.reset()
+        self._robot.set_joint_effort_target(joint_effort, env_ids=env_ids)
+
+        self.step_sim_no_action()
+
+    def step_sim_no_action(self):
+        """Step the simulation without an action. Used for resets."""
+        self.scene.write_data_to_sim()
+        self.sim.step(render=True)
```

Comment:
would it make sense to have render=False if running headless/no rendering needed?